File: LeetCode/src/slidingwindow/MaximumSumOfDistinctSubarraysWithLengthK.java
package slidingwindow;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class MaximumSumOfDistinctSubarraysWithLengthK {
    public long maximumSubarraySum(int[] nums, int k) {
        return maximumSubarraySumByHashMap(nums, k);
        // return maximumSubarraySumByHashSet(nums, k);
    }

    private long maximumSubarraySumByHashMap(int[] nums, int k) {
        long result = 0;
        Map<Integer, Integer> map = new HashMap<>();
        int left = 0, right = 0;
        long sum = 0;

        while (right < nums.length) {
            map.put(nums[right], map.getOrDefault(nums[right], 0) + 1);
            sum += nums[right];

            while (map.get(nums[right]) > 1) {
                map.put(nums[left], map.get(nums[left]) - 1);
                sum -= nums[left];
                left++;
            }

            if (right - left + 1 == k) {
                result = Math.max(result, sum);
                map.put(nums[left], map.get(nums[left]) - 1);
                sum -= nums[left];
                left++;
            }

            right++;
        }

        return result;
    }

    private long maximumSubarraySumByHashSet(int[] nums, int k) {
        long result = 0;
        Set<Integer> set = new HashSet<>();
        int left = 0, right = 0;
        long sum = 0;

        while (right < nums.length) {
            while (set.contains(nums[right])) {
                set.remove(nums[left]);
                sum -= nums[left];
                left++;
            }

            set.add(nums[right]);
            sum += nums[right];
            right++;

            if (right - left == k) {
                result = Math.max(result, sum);
                set.remove(nums[left]);
                sum -= nums[left];
                left++;
            }
        }

        return result;
    }

    public static void main(String[] args) {
        MaximumSumOfDistinctSubarraysWithLengthK solution = new MaximumSumOfDistinctSubarraysWithLengthK();
        int[] nums1 = {1, 2, 1, 3, 4, 3};
        int k1 = 3;
        System.out.println(solution.maximumSubarraySum(nums1, k1)); // Output: 8

        int[] nums2 = {1, 2, 1, 2, 3};
        int k2 = 2;
        System.out.println(solution.maximumSubarraySum(nums2, k2)); // Output: 5
    }
}

File: LeetCode_Python/slidingwindow/MaximumSumOfDistinctSubarraysWithLengthK.py
class MaximumSumOfDistinctSubarraysWithLengthK:
    def maximumSubarraySum(self, nums: list[int], k: int) -> int:
        n = len(nums)
        if n < k:
            return 0

        max_sum = 0
        current_sum = 0
        freq = {}
        left = 0

        for right in range(n):
            current_sum += nums[right]
            freq[nums[right]] = freq.get(nums[right], 0) + 1

            while freq[nums[right]] > 1:
                current_sum -= nums[left]
                freq[nums[left]] -= 1
                if freq[nums[left]] == 0:
                    del freq[nums[left]]
                left += 1

            if right - left + 1 == k:
                max_sum = max(max_sum, current_sum)
                current_sum -= nums[left]
                freq[nums[left]] -= 1
                if freq[nums[left]] == 0:
                    del freq[nums[left]]
                left += 1

        return max_sum


# Time Complexity: O(n)
# Space Complexity: O(n)

if __name__ == "__main__":
    solution = MaximumSumOfDistinctSubarraysWithLengthK()

    print("Expected Output: 15")
    print("Actual Output:", solution.maximumSubarraySum([1, 5, 4, 2, 9, 9, 9], 3))

    print("Expected Output: 0")
    print("Actual Output:", solution.maximumSubarraySum([4, 4, 4], 3))
